backend/src/controllers/v1/authController.ts

```diff
@@ -267,3 +301,753 @@ export const getNewToken = async (req: Request, res: Response) => {
+
+    const decodedToken = <jwt.IdentityAccessTokenJwtPayload>(
+        jwt.verify(accessToken, await getAuthSecret())
+    );
+
+    if (decodedToken.authTokenType !== AuthTokenType.IDENTITY_ACCESS_TOKEN) throw UnauthorizedRequestError();
+
+    const identityAccessToken = await IdentityAccessToken.findOne({
+        _id: decodedToken.identityAccessTokenId,
+        isAccessTokenRevoked: false
+    });
+
+    if (!identityAccessToken) throw UnauthorizedRequestError();
+
+    const {
+        accessTokenTTL,
+        accessTokenLastRenewedAt,
+        accessTokenMaxTTL,
+        createdAt: accessTokenCreatedAt
+    } = identityAccessToken;
+
+    if (accessTokenTTL === accessTokenMaxTTL) throw UnauthorizedRequestError({
+        message: "Failed to renew non-renewable access token"
+    });
+
+    // ttl check
+    if (accessTokenTTL > 0) {
+        const currentDate = new Date();
+        if (accessTokenLastRenewedAt) {
+            // access token has been renewed
+            const accessTokenRenewed = new Date(accessTokenLastRenewedAt);
+            const ttlInMilliseconds = accessTokenTTL * 1000;
+            const expirationDate = new Date(accessTokenRenewed.getTime() + ttlInMilliseconds);
+
+            if (currentDate > expirationDate) throw UnauthorizedRequestError({
+                message: "Failed to renew MI access token due to TTL expiration"
+            });
+        } else {
+            // access token has never been renewed
+            const accessTokenCreated = new Date(accessTokenCreatedAt);
+            const ttlInMilliseconds = accessTokenTTL * 1000;
+            const expirationDate = new Date(accessTokenCreated.getTime() + ttlInMilliseconds);
+
+            if (currentDate > expirationDate) throw UnauthorizedRequestError({
+                message: "Failed to renew MI access token due to TTL expiration"
+            });
+        }
+    }
+
+    // max ttl checks
+    if (accessTokenMaxTTL > 0) {
+        const accessTokenCreated = new Date(accessTokenCreatedAt);
+        const ttlInMilliseconds = accessTokenMaxTTL * 1000;
+        const currentDate = new Date();
+        const expirationDate = new Date(accessTokenCreated.getTime() + ttlInMilliseconds);
+
+        if (currentDate > expirationDate) throw UnauthorizedRequestError({
+            message: "Failed to renew MI access token due to Max TTL expiration"
+        });
+
+        const extendToDate = new Date(currentDate.getTime() + accessTokenTTL);
+        if (extendToDate > expirationDate) throw UnauthorizedRequestError({
+            message: "Failed to renew MI access token past its Max TTL expiration"
+        });
+    }
+
+    await IdentityAccessToken.findByIdAndUpdate(
+        identityAccessToken._id,
+        {
+            accessTokenLastRenewedAt: new Date()
+        }
+    );
+
+    return res.status(200).send({
+        accessToken,
+        expiresIn: identityAccessToken.accessTokenTTL,
```

Comment:
1. The response structure is meant to mirror OAuth2 where `expiresIn` is typically given in seconds.
2. Yes, this is taken into account in the `extendToDate` check; we check that if the current time + the access token TTL is greater than the expiration date (date created + max TTL) then we throw an error.

Reply:
Ah i misread it, milliseconds is only for when it's string without unit 

> Eg: 60, "2 days", "10h", "7d". A numeric value is interpreted as a seconds count. If you use a string be sure you provide the time units (days, hours, etc), otherwise milliseconds unit is used by default ("120" is equal to "120ms").